Source/WTF/wtf/glib/GThreadSafeWeakPtr.h

```diff
@@ -0,0 +1,89 @@
+/*
+ *  Copyright (C) 2024 Igalia S.L.
+ *
+ *  This library is free software; you can redistribute it and/or
+ *  modify it under the terms of the GNU Library General Public
+ *  License as published by the Free Software Foundation; either
+ *  version 2 of the License, or (at your option) any later version.
+ *
+ *  This library is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ *  Library General Public License for more details.
+ *
+ *  You should have received a copy of the GNU Library General Public License
+ *  along with this library; see the file COPYING.LIB.  If not, write to
+ *  the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ *  Boston, MA 02110-1301, USA.
+ *
+ */
+
+#pragma once
+
+#if USE(GLIB)
+
+#include "GRefPtr.h"
+
+#include <glib-object.h>
+#include <wtf/Noncopyable.h>
+
+namespace WTF {
+
+template <typename T> class GThreadSafeWeakPtr {
+    WTF_MAKE_NONCOPYABLE(GThreadSafeWeakPtr);
+public:
+    GThreadSafeWeakPtr()
+    {
+        g_weak_ref_init(&m_ref, nullptr);
+    }
+
+    explicit GThreadSafeWeakPtr(T* ptr)
+    {
+        RELEASE_ASSERT(!ptr || G_IS_OBJECT(ptr));
+        g_weak_ref_init(&m_ref, ptr);
+    }
+
+    GThreadSafeWeakPtr(GThreadSafeWeakPtr&& other)
+    {
+        auto strongRef = other.get();
+        g_weak_ref_set(&other.m_ref, nullptr);
+        g_weak_ref_init(&m_ref, strongRef.get());
+    }
+
+    ~GThreadSafeWeakPtr()
+    {
+        g_weak_ref_clear(&m_ref);
+    }
+
+    WARN_UNUSED_RETURN GRefPtr<T> get()
+    {
+        return adoptGRef(reinterpret_cast<T*>(g_weak_ref_get(&m_ref)));
+    }
+
+    void reset(T* ptr = nullptr)
+    {
+        RELEASE_ASSERT(!ptr || G_IS_OBJECT(ptr));
+        g_weak_ref_set(&m_ref, ptr);
+    }
+
+    GThreadSafeWeakPtr& operator=(std::nullptr_t)
+    {
+        reset();
+        return *this;
+    }
+
+    bool operator!()
+    {
+        auto strongRef = get();
+        return !strongRef;
+    }
```

Comment:
I'm not sure this is a good idea in this particular case because get is not cheap, so doing something like:

```
if (foo)
    bar(foo.get());
```

means we take and release the mutex twice and create a strong reference twice.

Reply:
Yeah I had the same thoughts... that's why i didn't use this operator in the patch. I can remove it maybe?

Reply:
Eh, it's debatable. I think the normal pattern here will be to get a strong ref and then do something with it if you succeed. Checking whether the weak pointer is still valid and then doing something would be weird because it could be invalidated between your check and your doing something.  So, maybe it would be best to not provide any operator!

Reply:
Yes, I think it's better to not provide operator!